src/scout_apm/falcon.py

```diff
@@ -0,0 +1,90 @@
+# coding=utf-8
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+from scout_apm.api import install
+from scout_apm.core.ignore import ignore_path
+from scout_apm.core.tracked_request import TrackedRequest
+
+try:
+    import falcon
+except ImportError:  # pragma: no cover
+    falcon = None
+
+# Falcon Middleware docs:
+# https://falcon.readthedocs.io/en/stable/api/middleware.html
+
+
+class ScoutMiddleware(object):
+    """
+    Falcon Middleware for integration with Scout APMi.
+    """
+
+    def __init__(self, config):
+        install(config=config)
+        self.api = None
+
+    def set_api(self, api):
+        if not isinstance(api, falcon.API):
+            raise ValueError("api should be an instance of falcon.API")
+        self.api = api
+
+    def process_request(self, req, resp):
+        if self.api is None:
+            raise RuntimeError(
+                "Call {}.set_api() before requests begin".format(
+                    self.__class__.__name__
+                )
+            )
+
+        tracked_request = TrackedRequest.instance()
+        tracked_request.mark_real_request()
+        req.context.scout_tracked_request = tracked_request
+
+        tracked_request.tag("path", req.path)
+        if ignore_path(req.path):
+            tracked_request.tag("ignore_transaction", True)
+
+        # Determine a remote IP to associate with the request. The value is
+        # spoofable by the requester so this is not suitable to use in any
+        # security sensitive context.
+        user_ip = (
+            req.get_header("x-forwarded-for", default="").split(",")[0]
+            or req.get_header("client-ip", default="").split(",")[0]
+            or req.remote_addr
+        )
+        tracked_request.tag("user_ip", user_ip)
+
+    def process_resource(self, req, resp, resource, params):
+        tracked_request = getattr(req.context, "scout_tracked_request", None)
+        if tracked_request is None:
+            # Somehow we didn't start a request - this might occur in
+            # combination with a pretty adversarial application, so guard
+            # against it, although if a request was started and the context was
+            # lost, other problems might occur.
+            return
+
+        # Find the current responder's name. Falcon passes middleware the
+        # current resource but unfortunately not the method being called, hence
+        # we have to go through routing again.
+        responder, _params, _resource, _uri_template = self.api._get_responder(req)
+        operation = "Controller/{}.{}.{}".format(
```

Comment:
What would an example of this look like? Are we going to have a high cardinality of `operation` names resulting from this?

Reply:
I don't think so. You'll see in the tests e.g. `Controller/tests.integration.test_falcon.HomeResource.on_get` and `Controller/tests.integration.test_falcon.HomeResource.on_get_suffixed`. As I understand from the Falcon docs, suffixed methods are the normal way of writing multiple types of response for a single underlying resource, so I think it's best to include them in the operation name.